Fixes inconsistent parentheses in power

Ryand1234 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
Fixes #6975
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed


#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES --> 
* printing 
   * fixed the parenthesization of powers in the latex printer
<!-- END RELEASE NOTES -->

Ryand1234 commented:
Can someone tell me why test is failing?

asmeurer commented:
There is some issue in the build for the pdf docs, which uses LaTeX. I wonder if `latex` is used somewhere to generate part of the documentation.

asmeurer commented on sympy/printing/latex.py:
```diff
@@ -616,22 +616,31 @@ def _print_Pow(self, expr):
         base = self.parenthesize(expr.base, PRECEDENCE['Pow'])
-        if '^' in base and expr.base.is_Symbol:
-            base = r"\left(%s\right)" % base
+        if '^' in base :
```
Checking for substrings doesn't seem very robust. What does the logic for pprint look like?

Ryand1234 replied:
What i understand is that first the expression is parenthesized if necessary then the side brackets **'|'** are added to the picture of expression. And then variables are filled in those brackets.

Ryand1234 replied:
Actually i manipulated the previous written logic to get the parentheses around necessary expression.

Ryand1234 commented:
Ping @asmeurer

Sc0rpi0n101 commented:
I don't think this fix needs to be included in the release notes. 

Also, can you squash your commits into one with a little description of what was actually changed.

asmeurer commented:
It can be included in the release notes, but make it relevant to the change for the end-user (like "fixed the parenthesization of powers in the latex printer").

Commit:
add Line.bisectors

sylee957 committed:
Use gaussian elimination for positive-definite check

sylee957 committed:
Resolve merge conflict

smitgajjar committed:
parsing(c_parser): added support for binary operators

- Added support for parsing binary operators `+`, `-`, `*`, `/`, `%`, `==`, `!=`, `<`, `<=`, `>`, `>=`, `&&` and `||`
- Added support for parsing assignment(`=`) from one variable to another
- Added support for assignment of integer as well as floating point literal to a variable
- Added support for an rhs expression comprising of any combination of the above operators
- Any form of parenthesised rhs expression will also be parsed successfully
- Added support for all forms of corresponding variable declaration as well as assignment
- Added support for recognizing boolean literals(`true` and `false`) in variable declaration as well as assignment
- Added support for boolean data type
- Corresponding tests are added

oscarbenjamin commented:
This has merge conflicts now.

jlherren committed:
Add assumption handlers for OneMatrix

jlherren committed:
Improve exceptions raised for matrices

- Raise more specific matrix exceptions whenever possible
- Raise correct exceptions on matrix inversion

jlherren committed:
Improve inversion of 2x2 block matrices

It previously tried to apply the formula even when A wasn't square.  It now
also considers invertibility of matrices to choose which formula to apply.

smichr committed:
int not Bool so Eq(bool, int) -> False

mijo2 committed:
solvers: Fixed some bugs in test_systems.py

There were some bugs in test_systems.py that were fixed in this commit.
This commit may or may not be squashed later.

mijo2 committed:
solvers: Removed the solver Linear, 3 equations, Order 1, Type 1

Removed the solver and ran tests to confirm that the new method
is computing the solution properly.

mijo2 committed:
solvers: Removing the solver Linear, 3 equations, Order 1, Type 2

Commit:
perm_groups.py: Add Coset class

Commit:
removed some changes

Commit:
changes example

asmeurer committed:
Merge pull request #19090 from sylee957/sphinx_deprecation

Change deprecated sphinx from sympylive

Commit:
added check for dir

rj-calvin committed:
added test case for updated replace_with_arrays

smichr committed:
S.true != 1, too

jlherren committed:
Untangle ShapeError imports to be able to remove unused import

jlherren committed:
Use sympify in BlockDiagMatrix constructor

Smit-create committed:
test for labels

sylee957 committed:
Merge pull request #19093 from jlherren/blockmatrix

Various block matrix improvements

Commit:
perm_groups.py : Add relations in is_subgroup and __contains__ in SymmetricPermutationGroup

Commit:
add more tests and some minor improvements

smichr committed:
Merge pull request #19094 from smichr/linbis

add Line.bisectors

Commit:
added examples

Commit:
perm_groups.py : minor improvements

smichr committed:
Merge pull request #19054 from smichr/solvebool

be more specific when testing for T/F

Commit:
Merge pull request #18993 from mohitacecode/issue_18804

Solves key error in creating polycyclic group, Improves _sift, induced_pcgs

jlherren committed:
Fix evaluation of inverse of inverse

Inverse(Inverse(A)).doit() was previously returning A ** 1.  This also moves
the responsibility of inverting a MatPow to the MatPow._eval_power method
instead of the Inverse class.

sylee957 committed:
Merge pull request #19079 from Smit-create/plot_label

Accept label as argument in plotting

oscarbenjamin committed:
Merge pull request #18877 from smitgajjar/py-upgrade2

Upgrading to Python3 (2) (pyupgrade to files from benchmarks to crypto)

sylee957 committed:
Use numpydoc format for plot

oscarbenjamin committed:
Merge pull request #19101 from sylee957/plot_doc

Improve plot docs

jlherren committed:
Forbid most powers of non-square matrices and code clean-up

This delegates special MatPow behavior to the respective classes instead of
handling them inside MatPow.doit().

ebraminio committed:
Remove two minor references to py2's print

sylee957 committed:
Merge pull request #19105 from ebraminio/minor-py2-rm

Remove two minor references to py2's print

jlherren committed:
Stabilize two flaky tests: test_pmint_erf/test_V12

They randomly failed due to itermonomials() yielding values in
non-deterministic order, because it uses set().

jlherren committed:
Fix combine_kronecker() manipulation for non-square matrices

(A x B) ** m is no longer transformed to A ** m x B ** m for non-square
matrices A, B, where x is the kronecker product.

tacaswell committed:
FIX: don't try to autoscale 3D axes

tacaswell committed:
MNT: Only autoscale if not manually set

jksuom committed:
Merge pull request #19087 from moorepants/mpl-less-than-3.2

Use matplotlib < 3.2 on Travis.

oscarbenjamin committed:
Merge pull request #19108 from jlherren/fix-unstable-tests

Stabilize two flaky tests: test_pmint_erf/test_V12

jlherren committed:
Raise nicer error when calling as_explicit() on symbolic-shaped matrix

mijo2 committed:
doc: Updated the file ode.rst

Removed the documentation details of the solvers that are to be removed
after adding the first general case solver

Commit:
added input checks for dir

mijo2 committed:
solvers: Updated the code

Some number of minor updates in the code were made.

Mohitbalwani26 committed:
refactored nth_algebraic examples and fixed a small bug in separable_reduced hint

Mohitbalwani26 committed:
fixed unused import

sbt4104 committed:
ZeroMatrix should not return False.

certik committed:
Doc: update the GSoC reports

asmeurer committed:
Merge pull request #19111 from certik/gsoc

Doc: update the GSoC reports

Mohitbalwani26 committed:
reverted back the fix for separable_reduced as it failed for doctest

sylee957 committed:
Merge pull request #19110 from sbt4104/master

ZeroMatrix should not return False.

mijo2 committed:
solvers: Added regression tests for issue #15474

mijo2 committed:
solvers: Moved test cases

Moved the test cases for the systems of ODEs in test_ode.py that are
now covered by the new general case solver.

mijo2 committed:
solvers: Updated the _get_func_order function

Updated the logic of _get_func_order function and minimised the
implmentation length

mijo2 committed:
solvers: Refactored linear_eq_to_matrix in systems.py

linear_eq_to_matrix function is used in systems.py and hence it is
refactored.

jlherren committed:
Completely disallow powers of non-square matrices

mijo2 committed:
solvers: Edited the import names

Mohitbalwani26 committed:
making one function call per line and fixed gen argument

jlherren committed:
Remove unnecessary method override

jlherren committed:
Fix bc_block_plus_ident() losing unrelated summands

smichr committed:
19113: public roots require symbol generator

The issue of solving cos(x)**5 - sin(x)**5 is circumvented
by a better solve strategy that rewrite sin/cos as tan via
TR2i. But the issue of roots being generated for expressions
with non-symbol generators is adjusted on two fronts:

1) public functions that do root-related work reject expressions
with non-symbol generators
2) univariate Poly which find roots will always return a canonical
RootOf with generator `x`.

sylee957 committed:
Merge pull request #19116 from jlherren/fix-bc_block_plus_ident

Fix bc_block_plus_ident() losing unrelated summands

RituRajSingh878 committed:
added new reduced_separable method in ode

Mohitbalwani26 committed:
reverted back the fix in ode.py as new PR was open for that

asmeurer committed:
Merge pull request #19006 from sylee957/sort_key_sympify

Deprecate sympify string fallback

Smit-create committed:
allow custom distributions

Smit-create committed:
fix failing tests

mijo2 committed:
solvers: Added linear_ode_to_matrix function

mijo2 committed:
solvers: Added regression test case for issue #15474

mijo2 committed:
solvers: Docstring fix

A bug was fixed in the dcocstring of the function linear_ode_to_matrix

mijo2 committed:
solvers: Added test cases for linear_ode_to_matrix

Smit-create committed:
tests and docs

RituRajSingh878 committed:
added test and modified code

RituRajSingh878 committed:
removed invalid syntax

jlherren committed:
Move code out of try..except block

Mohitbalwani26 committed:
added a check whether solution returned by dsolve without passing hint is correct

jlherren committed:
MatMul: Add and improve tests, properly call parent constructor

sylee957 committed:
PEP8 fixes

RituRajSingh878 committed:
updated test

sylee957 committed:
Merge pull request #19077 from mohitacecode/Coset_class

perm_groups.py: Add Coset class

sylee957 committed:
Merge pull request #19104 from jlherren/matpower

Clean up of MatPow code, forbid non-square matrix powers

sylee957 committed:
Merge pull request #19083 from jlherren/fix-cse-opts

Fix CSE to not construct invalid tree

mijo2 committed:
solvers: Updated the docstring of matrix_exp function

Updated the documentation of matrix_exp function in systems.py
according to the Sympy's documentation style guidelines

smichr committed:
Merge pull request #19117 from smichr/19113

19113: public roots require symbol generator

Commit:
add a Rule

Add this rule:  'Pochhammer[x,y]':'rf(x,y)'

Commit:
add Rules

add 'ArcTan[x,y]':'atan2(y,x)'

Sc0rpi0n101 committed:
Merge pull request #19029 from smitgajjar/binary_op

parsing(c_parser): added full support for Binary Operators and any type of complicated Assignment in C parser

smichr committed:
enhance Mul.is_integer

RituRajSingh878 committed:
added checkodesol in a test

Mohitbalwani26 committed:
cleaning the code of newly added internal test functions

iammosespaulr committed:
Update cse_opts.py

Fixes #19136

mijo2 committed:
solvers: Updated docstring for function matrix_exp_jordan_form

mijo2 committed:
solvers: Small changes in some docstrings

oscarbenjamin committed:
Merge pull request #19122 from RituRajSingh878/reduced_ode

added new reduced_separable method in ode

oscarbenjamin committed:
Merge pull request #19063 from Mohitbalwani26/refactoring_tests

refactor(test_ode): Factorable examples

Commit:
Add test

Add test for mathematica paser.

Commit:
perm_groups and permutation : modify __mul__

Commit:
PEP8 fix

smichr committed:
Update cse_opts.py

smichr committed:
Apply suggestions from code review

smichr committed:
Merge pull request #19137 from iammosespaulr/patch-3

Update cse_opts.py

smitgajjar committed:
parsing(c_parser): added support for unary ops and paren expr
- Added support for unary operators:
    - unary plus(+), unary minus(-), pre/post increment(++), pre/post decrement(--) and logical NOT(!)
- Added support for parenthesized expression
- Added corresponding tests
    - Added tests for unary operators
    - Added tests for parenthesized expression
    - Added tests for Exceptions raised in parsing binary operators, variable declarations and unary operators
- Improved error messages of Exceptions raised in parsing binary operators, variable declarations and unary operators

Other notes:
- Dereferencing operator(*), Address operator(&) and bitwise NOT(~) are the unary operators that have not been implemented because of unavailability of such nodes or classes in SymPy
- If unary operators are the children of binary operator in AST, it is not possible to parse yet!

sylee957 committed:
Merge pull request #19139 from mohitacecode/modifymul

Define multiplication between Permutation and PermutationGroup

Commit:
delete __name__==__main__ ,instead put it outside

delete __name__==__main__

Commit:
delete __name__=="__main__"

delete __name__=="__main__"  ,instead, put it in test_mathematica.py

Commit:
Add test example

Add 'ArcTan[x,y]': 'atan2(y,x)' and  'Pochhammer[x,y]': 'rf(x,y)'.

jlherren committed:
PEP 8 fix

Co-Authored-By: Smit Gajjar <38134283+smitgajjar@users.noreply.github.com>

sylee957 committed:
Correct grammar

smichr committed:
Merge pull request #19060 from Psycho-Pirate/19059

Fixed error with Pyglet Plotting

smichr committed:
Merge pull request #19142 from jlherren/as-explicit

Raise nicer error when calling as_explicit() on symbolic-shaped matrix

sylee957 committed:
Merge pull request #19135 from wuyudi/master

add 2 rules in mathematica parser

akirakyle committed:
add _eval_adjoint and _eval_transpose to ExpBase

mijo2 committed:
solvers: Added NonLinearError in solveset

mijo2 committed:
solvers: Added ODENonLinearError in systems.py

mijo2 committed:
solvers: Updated a doctest in linsolve

A doctest in function linsolve was giving incorrect results. It was
updated.

mijo2 committed:
solvers: Added canonical equations function

iammosespaulr committed:
add support for binom in latex

jlherren committed:
Improve printing of matrix slices

iammosespaulr committed:
fixes

sylee957 committed:
Merge pull request #19151 from akirakyle/master

add _eval_adjoint and _eval_transpose to ExpBase

smichr committed:
enhance Pow.is_integer

smichr committed:
add assumption helpers; relocate

smichr committed:
flake8 mods to Mul

smichr committed:
add note about storage of assumptions

mijo2 committed:
solvers: Updated the name of NonLinearError to NonlinearError

mijo2 committed:
solvers: Updated canonical_equations function

Made canonical_equations private.

mijo2 committed:
solvers: Updated the solver

The solver can now collect exp terms together.

iammosespaulr committed:
downgrade to 4.7.1

mijo2 committed:
solvers: Added regression test case for issue #19150

Upabjojr committed:
Merge pull request #19091 from cjayross/tensor_contraction

Fixed tensor contraction problem

jlherren committed:
Improve 2x2 block matrix inversion by implementing all formulas

sylee957 committed:
Merge pull request #19147 from iammosespaulr/origin/latex

add support for binom in latex

smichr committed:
Merge pull request #19155 from smichr/assumptions

adding to and moving assumptions

smitgajjar committed:
ran pyupgrade across files ranging from deprecated to ntheory module

smitgajjar committed:
added coding utf-8 to resolve travis fail

smitgajjar committed:
added coding utf-8

smitgajjar committed:
trigger travis

smichr committed:
19157 recognize when quintics fails

smichr committed:
replace nonsym collectables with symbols

mijo2 committed:
solvers: Updated some of the test cases in test_ode.py

Updated the test cases and added a required comment for future
reference

mijo2 committed:
doc: Updated the path for the general case solver

mijo2 committed:
solvers: Updated the incomplete documentation in function linear_ode_to_matrix

mijo2 committed:
solvers: Added some minor changes requested.

mijo2 committed:
solvers: Added a workaround for the collect function

Collect function doesn't work correctly for some cases, (more information
in issue #19149). A workaround is added in our general case solver.

mijo2 committed:
doc: Added documentation details for some functions in systems.py

Added documentation details for new functions like linear_ode_to_matrix,
matrix_exp_jordan_form, matrix_exp.

mijo2 committed:
solvers: Updated doc files for successful build.

smichr committed:
Merge pull request #19165 from smichr/19149

collecting nonsymbols; solving quintics

oscarbenjamin committed:
Merge pull request #19160 from smitgajjar/pyupgrade3

Upgrading to python3 (3)

sylee957 committed:
Fix numpy printing of unevaluated pow

mijo2 committed:
Merge branch 'master' into 15549_matrix_exp

mijo2 committed:
solvers: Removing the workaround in systems.py

oscarbenjamin committed:
Merge pull request #19170 from sylee957/fix_numpy_pow

Fix numpy printing of unevaluated pow

oscarbenjamin committed:
Merge pull request #18720 from mijo2/15549_matrix_exp

Using matrix exp for solving linear differential equations

Mohitbalwani26 committed:
refactoring riccati_special_minus2 test

oscarbenjamin committed:
Merge pull request #19176 from Mohitbalwani26/refactoring_tests

refactor(test_ode): Riccati_special_minus2

smichr committed:
Merge pull request #8348 from cbm755/not_all_basic

Basic: docstring: not all objects in SymPy are Basic

MartinThoma committed:
DOC: Use README.rst for pypi.org page

iammosespaulr committed:
Okay!

Mohitbalwani26 committed:
refactor(test_ode.py): regression tests

Mohitbalwani26 committed:
marked examples as xfail for the hints which they were failing

MartinThoma committed:
Convert README to markdown with pandoc

czgdp1807 committed:
Merge pull request #19123 from Smit-create/custom_dists

Allow custom probability distributions

czgdp1807 committed:
removed user guide

Mohitbalwani26 committed:
added _test_dummy_sol function to handle solutions with dummy variable and shifted most of try block outside as it had bugs

Mohitbalwani26 committed:
import fix

sylee957 committed:
Change plot_parametric docs to numpydoc format

oscarbenjamin committed:
Merge pull request #19179 from MartinThoma/pypi

DOC: Use README.rst for pypi.org page

MartinThoma committed:
STY: Use Python 3 syntax in vector module

sylee957 committed:
Improve the notes

sylee957 committed:
Add plot example for notes

iammosespaulr committed:
upgrade Mul.is_integer and fix tests

Co-authored-by: Chris Smith <smichr@gmail.com>

fix failing

czgdp1807 committed:
removed unwanted pics

oscarbenjamin committed:
Merge pull request #19194 from MartinThoma/vector-py3

Use Python 3 syntax in vector module

smichr committed:
Merge pull request #19182 from iammosespaulr/assump

fix Mul.is_integer

jlherren committed:
Always print matrix slices in a way to not be confused with matrix elements

sylee957 committed:
Do not attempt to factor if the expression has float

sylee957 committed:
Update sympy/plotting/plot.py

Co-Authored-By: Christopher Smith <smichr@gmail.com>

oscarbenjamin committed:
Merge pull request #18985 from sympy/revert-18434-17119

Revert "Handling Float Integrals"

sylee957 committed:
Merge pull request #19201 from jlherren/slice-printing

Improve matrix slice printing

sylee957 committed:
Fix flake8

sylee957 committed:
Improve docs for xlim and ylim

sylee957 committed:
Improve example description

sylee957 committed:
Merge branch 'improve_plot_docs' of https://github.com/sylee957/sympy into improve_plot_docs

sylee957 committed:
Fix typo

sylee957 committed:
Merge pull request #19088 from tacaswell/fix_autoscale

FIX: ensure autoscale_view gets called

sylee957 committed:
Remove division by 2 for generalized cases

sylee957 committed:
Use Gershgorin circle theorem for testing positive definite

sylee957 committed:
Fix wrong absolute value computation

sylee957 committed:
Fix typo

sylee957 committed:
Fix typo

sylee957 committed:
Return false for nonsquare cases

asmeurer commented on sympy/printing/latex.py:
```diff
@@ -616,22 +616,35 @@ def _print_Pow(self, expr):
+        else:
+            exp = r"{%s}" % exp
+
+        ''' checks if base expression contains '^' and it is not considered
```
Use # for comments.

sylee957 committed:
Return false when diagonals are strictly negative

sylee957 committed:
Fix wrong stats test cases

aktech committed:
Update all LICENSE Copyright year

sylee957 committed:
Fix invalid latex control

sylee957 committed:
Change index notation

Commit:
Merge pull request #19213 from sympy/license-2020

Update all LICENSE Copyright year

sylee957 committed:
Use more cheaper test for solvers

sylee957 committed:
Add regression test for #18955

Mohitbalwani26 committed:
simplified the code for _test_paticular_example_

Mohitbalwani26 committed:
_test_an_example function calls test_particular_example for the code output

jlherren committed:
Remove unused import

eric-wieser committed:
Speed up Symbol.sort_key

There's no need to go through `StrPrinter` here.

eric-wieser committed:
Speed up Dummy.sort_key

There's no need to go through `StrPrinter` here.

MartinThoma committed:
STY: Remove trailing whitespace

asmeurer committed:
Merge pull request #19190 from czgdp1807/user_guide

User guide removed

smichr committed:
Merge pull request #19216 from eric-wieser/patch-5

Speed up Symbol.sort_key

aktech committed:
Make copyright in docs dynamic

Mohitbalwani26 committed:
refactored liouville test examples

Sc0rpi0n101 committed:
Merge pull request #19140 from smitgajjar/unary_op

parsing(c_parser): added support for unary operators and outermost parenthesized expression

asmeurer committed:
Put the doctr environment variable only in the build that needs it

Aside from being more secure, this fixes the environment variable listing on
the Travis site from showing DOCTR_DEPLOY_ENCRYPTION_KEY=[secure] as the first
thing for every build.

asmeurer committed:
Order the environment variables better in .travis.yml

The environment variables are shown on the build listing on the Travis site in
the order they are in the file, so it's best to put them in a consistent order
with the most important variables first.

sylee957 committed:
Merge pull request #19226 from asmeurer/doctr-env

Put the doctr environment variable only in the build that needs it

sylee957 committed:
Merge pull request #19219 from aktech/dynamic-copyright

Make copyright in docs dynamic

sylee957 committed:
Inherit MatrixExpr for immutable sparse matrix

sylee957 committed:
Add tests

sylee957 committed:
Fix test

oscarbenjamin committed:
Merge pull request #19205 from sylee957/improve_positive_definite

Use gaussian elimination for positive-definite check

oscarbenjamin committed:
Merge pull request #19228 from sylee957/fix_immutable_sparse

Inherit MatrixExpr for ImmutableSparseMatrix

sylee957 committed:
Fix creation of matrix with list of ndarray

sylee957 committed:
Revert redundant change

sylee957 committed:
Fix test without numpy

oscarbenjamin committed:
Merge pull request #19229 from sylee957/fix_ndarray_matrix

Fix creation of matrix with list of ndarray

oscarbenjamin committed:
Remove expand_mul call in cholesky and LDLdecompisition

Both cholesky and LDLdecompisition used expand_mul to simplify some
calculations but this can lead to significant slowdowns in some cases.
This commit removes the calls to expand_mul from each of these
functions. A few tests needed to be changed because the output is no
longer simplified to the same extent.

See https://github.com/sympy/sympy/issues/19200

sylee957 committed:
Merge pull request #19215 from jlherren/improve-2x2-block-inversion

Improve 2x2 block matrix inversion with off-diagonal invertible blocks

eric-wieser committed:
Respect `inv_trig_style` in inverse hyperbolic trig functions

Upstreamed from the galgebra printer, which seemed to want these.

iammosespaulr committed:
fix_holonomic caching error

oscarbenjamin committed:
Merge pull request #19184 from Mohitbalwani26/refatoring_tests

refactor(test_ode.py): regression tests

sylee957 committed:
Merge pull request #19235 from eric-wieser/include-hyperbolic-in-inv_trig

Respect `inv_trig_style` in inverse hyperbolic trig functions

sylee957 committed:
Display examples for plot_implicit

sylee957 committed:
Fix doctest

oscarbenjamin committed:
Merge pull request #19239 from sylee957/plot_implicit_examples

Display examples for plot_implicit

oscarbenjamin committed:
Merge pull request #19233 from oscarbenjamin/pr_expand_mul_revert

Remove expand_mul call in cholesky and LDLdecompisition

oscarbenjamin committed:
Bump version to 1.7.dev

smichr committed:
remove 2nd use of is

oscarbenjamin committed:
Merge pull request #19244 from oscarbenjamin/pr_bump_1.7.dev

Bump version to 1.7.dev

smichr committed:
Merge pull request #19192 from sylee957/improve_plot_docs

Improve plot_parametric docs

oscarbenjamin committed:
fix(testing): Set the warning filter around each test

When running tests/doctests with sympy's own test runners a warning
filter is used to turn DeprecationWarning/SymPyDeprecationWarning into
an error. The filter was only set one at the start of running the tests.
This means that if the filter is modified by one test then it will still
be modified when subsequent tests run.

This commit adds a raise_on_deprecated context manager in runtests.py
and uses it to set the warning filter in the test runners. The filter is
set immediately around running the test and is reset before each
individual test. This has picked up on a couple of failures that will
need to be fixed in subsequent commits.

oscarbenjamin committed:
fix(core): don't parse strings in Basic.has

The Basic.has method is changed from using sympify to _sympify so that
string input is no longer accepted and so that functions are no longer
parsed as strings.

A line in solveset needed to be fixed because it was erroneously calling
expr.as_independent([x, y]) rather than expr.as_independent(x, y).

BREAKING CHANGE

Basic.has no longer accepts strings so e.g. Symbol('x').has('x') will
raise an exception.

oscarbenjamin committed:
fix(printing): don't sympify strings in doctest

The tableform doctest relied on implicit sympification of strings into
symbols. This commit calls Symbol explicitly to remove the warning from
sympify.

oscarbenjamin committed:
fix(core): don't sympify strings in doctest

The Dict doctest relied on auto-sympification of strings in calling
D.has('one'). This commit uses Symbol explicitly because has no longer
accepts string input after 9c57fd06f76226507f4bc910d55c60de7213f4e9

BREAKING CHANGE: Using D.has('x') where D is a SymPy Dict will raise an
exception. Use D.has(Symbol('x')) instead.

oscarbenjamin committed:
Merge pull request #19247 from oscarbenjamin/pr_warnings_tests

Fix handling of warnings in the test runner

oscarbenjamin committed:
Merge pull request #19236 from iammosespaulr/fix_holonomic

fix_holonomic caching error

oscarbenjamin committed:
Merge pull request #19214 from MartinThoma/trailing-whitespace

STY: Remove trailing whitespace

czgdp1807 commented:
@Ryand1234 Are you still working on this PR? Please resolve the conflicts.

riyandhiman14 committed:
add parenthesis around base expression if expression in power contains expression with power.

riyandhiman14 committed:
add parenthesis around base expression if expression in power contains expression with power.

riyandhiman14 committed:
Merge branch 'coeff' of https://github.com/Ryand1234/sympy into coeff

Ryand1234 commented:
I'll create a new PR with new branch because there is some problem in this branch.
